### PR TITLE
Add protected routes

### PR DESCRIPTION
**Pull request by yash9263**

### CODE REVIEW

File: src/components/RequireAuth/index.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  const location = useLocation()
+
+  const [token] = useLocalStorage<{ encodedToken: string; userInfo: object }>('user_token')
+
+  if (!token || !token.userInfo) {
+    return <Navigate to="/login" state={{ from: location }} replace />
+  }
+  const { userInfo: user, encodedToken } = token
+
+  return <Outlet context={{ user, encodedToken }} />
```
Comment (sainath7878) [inline src/components/RequireAuth/index.tsx:15]: doubt: Is the context a prop here?

**Review by sainath7878 — APPROVED** [review]
Nothing to suggest.Everything looked awesome to me. In fact, I've learn't a lot reviewing your code. Your code, folder structure is awesome. Great work!!

File: .gitignore [hunk 2]
```diff
@@ -22,4 +22,5 @@
 npm-debug.log*
 yarn-debug.log*
 yarn-error.log*
-src/test.styled.tsx
\ No newline at end of file
+src/test.styled.tsx
```
Comment (govindappaarun) [inline .gitignore:25]: any reasons to add hook to ignore list

File: src/components/ProductCard/index.tsx [hunk 3]
```diff
@@ -83,4 +104,34 @@ export const WishlistBtn = styled(Button)`
   }
 `
 
+export const ActionBtn = styled(Button)`
+  border-radius: initial;
+  & > svg {
+  }
+`
+
+export const fadeIn = keyframes`
```
Comment (govindappaarun) [inline src/components/ProductCard/index.tsx:113]: this can go into utils

File: src/context/auth-context.tsx [hunk 4]
```diff
@@ -39,14 +72,16 @@ const useAuthProvider = () => {
...
     } catch (error) {
+      setUser(undefined)
       console.error(error)
     }
   }
 
   const logOut = async () => {
-    setUser(null)
+    setUser(undefined)
     if (typeof window !== 'undefined') window.localStorage.removeItem('user_token')
```
Comment (govindappaarun) [inline src/context/auth-context.tsx:85]: do you also want to remove user details here?

**Review by govindappaarun — APPROVED** [review]
I have few questions/suggestions, As you are still working on it, Would like to see once things are completed

File: src/components/ProductCard/index.tsx (hunk 3 above)
Comment (yash9263) [reply src/components/ProductCard/index.tsx:113]: The UI stuff is in progress. I am mostly focusing on the functionality now will refactor that in future

File: .gitignore (hunk 2 above)
Comment (yash9263) [reply .gitignore:25]: I am just experimenting with the styled-components APIs that's why added it git ignore

**Pull request merged**